docs/settings.rst
.. _qcpump-settings:

QCPump Settings
===============

You can find the location of your QCPump settings file location file by
launching QCPump and going to File->About:

.. figure:: pumps/dev/images/config_file_loc.png


You can edit the settings.json file with any text editor:

.. figure:: pumps/dev/images/config_file.png


Available Settings
------------------

You may add one or more of the following settings to the settings.json file
to override the default values:


BROWSER_USER_AGENT (string)
    The User-Agent to use when making outgoing web requests. (Default
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like
    Gecko) Chrome/70.0.3538.102 Safari/537.36 Edge/18.19582")

DB_CONNECT_TIMEOUT (integer)
    Timeout in seconds for database connections where available. (Default 30)

DEBUG (`true`, `false`)
    Currently only used to redirect std input / output. Must be set to `true`
    if you want to use an interactive debugger while developing QCPump.
    (Default: `false`)

LOG_LEVEL (`debug, info, warning, error, critical`)
    Choose the QCPump application logging level (individual pumps are not
    affected by this).  One of.
    (Default: `info`)

LOG_TO_CONSOLE (`true`, `false`):
    Should logs be written to console as well as log files?

PUMP_DIRECTORIES (list of file paths or null)
    Set to list of other directories to include user defined pump types from.
    See :ref:`pumps-developing`.

.. _qcpump-settings-startup:

PUMP_ON_STARTUP (`true`, `false`)
    Should QCPump immediately start pumping when it is launched.  This is useful
    for e.g. adding QCPump to a startup folder so it launches when a machine is rebooted
    and starts pumping immediately. (Default `false`)


docs/pumps/config.rst
.. _pumps-configure-new:

Configuring New Pumps
=====================

When you first launch QCPump you will be met with a more or less blank slate.

.. figure:: images/fresh.png
    :alt: QCPump with no pumps installed

    QCPump with no pumps installed


Click the *Add Pump* button at the top right. Select the *Pump Type* you want
to create, give the *Pump* a meaningful name and click *OK*.

.. figure:: images/add-new-pump.png
    :alt: QCPump with no pumps installed

    QCPump with no pumps installed


After adding you click *OK* you will see a tab for your new *Pump* displayed:

.. figure:: images/pump-added.png
    :alt: A new pump added

    A new pump added


On the left of the tab are the configuration options for the pump and on the
right is a logging window which is used to display messages from your pump.

The configuration panel is separated into different configuration sections.
All pumps have a common set of configuration options included under the *Pump
Configuration* heading which control things like whether this pump is active
and how often the pump should run.

.. _pumps-common-config:

Common Pump Configuration Options
---------------------------------

Every *Pump* you configure has the following options available in the *Pump
Configuration* section:

Type
    Displays the *Pump Type* of the current *Pump* (not editable)

Name
    Displays the name of the current *Pump* (not editable)

Interval (s)
    How often the pump should run in seconds.

Log Level
    A dropdown allowing you set the verbosity of logging for this Pump.  Set to *debug* to
    get the maximum level of verbosity.  Logging messages will be shown both in the
    status panel as well as written to log files.  The location of the log files can 
    be seen by accessing the *About* menu option in the *File* menu.

Active
    Select whether this pump will be activated when you click the *Run Pumps* button.


*Pump Type* specific options are covered in their own pages:


.. toctree::
    :maxdepth: 2
    :caption: Pump Type Specific Options:

    filemover.rst
    dqa3.rst
    mpc.rst



Configuration Validation
........................

Any time you edit a configuration value your *Pump* will be re-validated.  This
may include actions such as ensuring a file system directory exists, or sending
an HTTP request to check that a QATrack+ API endpoint is reachable.  

.. note::

    Currently the *Pump* tab is disabled during validation to prevent multiple
    validations from running simultaneously.

After validation, a message will be shown indicating whether each configuration
setting is valid or not. 

You can force a revalidation (say if some external factor was reconfigured /
updated) by pressing the `Revalidate` button.


Configuration Dependencies
..........................

Some *Pump Types* will have configuration sections which depend on other
configuration sections being validated successfully before they can be
completed.  For example, selecting a Unit configured in QATrack+ will not be 
possible until QCPump has verified it can talk to your QATrack+ API. 

When a dependency of a configuration section has not been met, the dependent
section will be disabled and a message will be shown to inform you of which
sections need to be completed before the dependent section can be verified.

.. figure:: images/validation-dependencies.png
    :alt: QCPump config with unmet dependencies

    QCPump config with unmet dependencies


Multiple Configuration Subsections
..................................

Some configuration sections allow you to use multiple repeated subsections.  An
example of this would be the configuration used by the DQA3 Pump Types to map
DQA3 Machines to QATrack+ Units:


.. figure:: images/multiple.png
    :alt: Multiple config subsections

    Multiple config subsections

To add new configuration subsections click the *+* button, and to remove a
subsection, highlight one of its configuration options and click the *-*.

.. _pumps-save:

Saving your Pumps
-----------------

Before you can run your pumps, they all must be saved (pumps which haven't been
saved will be marked with asterisks surrounding their pump tabs name). Once you
have configured and validated your *Pump* click the `Save` button at the top of
its config section. The location of the configuration files for QCPump can 
be found by accessing the *About* menu option in the *File* menu.

Reseting a Configuration
------------------------

If you want to reset your Pump config, click the `Reset` button at the top of
its config.  This will reset the pumps config to the most recently saved
version of the pumps configuration.

Deleting a Pump
---------------

If you want to delete a Pump config, click the `Delete` button at the top of
its config.  This will permanently delete this pump and remove its
configuration file from disk.




docs/release_notes.rst
.. _release-notes:

Release Notes
=============

v0.3.2
------

* Handle non-missing test 400 Bad Requests when uploading to QATrack+

v0.3.1
------

* The File Mover pump types can now be set to `Copy` mode to have source files
  copied to the destination directory rather than moved

* The Python package `python-certifi-win32` has been added so that requests
  can use the Windows Certificate Store for SSL verification rather than using
  its bundled certificate chain.  This should resolve some issues users were
  having with firewalls & network monitoring software.

* Added a missing permission for the DQA3 QCPump Firebird user

v0.3.0
------

General Features & Fixes
........................


* A bug has been fixed where only the last subsection of a *Multiple*
  configuration section was being validated.

* Pumps which are marked as *Inactive* will not run validation code until they
  are re-activated.  This eliminiates un-necessary network calls and other
  validation work which, in addition to being more efficient overall, makes
  debugging QCPump itself simpler when multiple pumps are configured.

* New MPC pump type for uploading MPC results. See :ref:`pump_type-mpc`.

* New generic pump types for uploading Text & Binary Files to QATrack+ have
  been implemented. See :ref:`pump_type-qatrack-upload`.

* A `DISPLAY_NAME` attribute has been added to Pump Types to aid with grouping
  together similar pump types when adding new pumps.

* Warning level debug messages were being logged as errors. This has been fixed.

* A new `PUMP_ON_STARTUP` (see :ref:`qcpump-settings`) setting has been added
  to allow pumping to begin immediately after QCPump is launched. This allows you
  to place QCPump in a startup folder and have it launched & start pumping when
  your computer is restarted.


DQA3 Pump Type Changes
......................

* The `DATEADD` for calculating a `work_completed` value in Firebird DQA3
  queries has been eliminated in order to allow the query to work with Firebird
  versions < 2.1.  `work_completed` is now just calculated in Python code
  instead.

* The template for looking up Test Lists for beams now defaults to: 
  
    .. code::

        Daily QA3 Results: {{ beam_name }}

  where `beam_name` is is the DQA3 test name (e.g. '6MeV', '6MV WDG', '6MV EDW
  60 Weekly', '20 MeV DQA3 Daily').  This allows QCPump to handle a wider variety
  of beam types/configurations.

* More context variables are available when generating your test list name.  In
  most cases you should only need to use `beam_name`, however other variables
  are available should you need them. See the :ref:`DQA3 Test List Name docs
  <pump_type-dqa3-test-list>`.


* New :ref:`Multiple Beam Per Test List <pump_type-dqa3-grouped>` DQA3 pumps
  have been added which will group results from multiple measurements together
  based on the results being recorded in a short window of time.  There are two
  disadvantages to using the Multiple Beams Per Test List:

    1. If you have many beams configured this will result in long test
       lists which can impact performance when uploading data, or reviewing
       data in QATrack+.

    2. If you perform a measurement twice (e.g. take 2 6X measurements), only
       the 2nd result will be included.

* QATrack+ Unit names will now be displayed along with their Site in order
  to disambiguate units with the same name

* DQA3 machine names will now be shown with their Room name to disambiguate
  machines using the same tree names.


qcpump/settings.py
import json
import os
from pathlib import Path
import sys

import appdirs

# first figure out where our application is running.  This will differ
# depending on platform and whether or not the application is frozen by
# py2exe/pyinstaller/cx_freeze/other
frozen = getattr(sys, 'frozen', '')
if not frozen:  # not frozen: in regular python interpreter
    root = os.path.abspath(os.path.dirname(__file__))
    frozen = False
elif frozen in ('dll', 'console_exe', 'windows_exe', 1):  # py2exe/pyinstaller:
    frozen = True
    try:
        # pyinstaller
        base_path = sys._MEIPASS
        root = os.path.dirname(base_path)
    except Exception:
        # cx_freeze
        root = os.path.dirname(sys.executable)


def resource_path(relative_path):
    """ Get absolute path to resource, works for dev and for PyInstaller """
    if frozen:
        return os.path.join(root, "qcpump", relative_path)
    return os.path.join(root, relative_path)


class Settings:
    """A class to hold QCPump Application settings.  Default settings can be
    overridden using a settings.json file"""

    ROOT = root  # where is the root directory
    FROZEN = frozen  # is the application "compiled" by py2exe/pyinstaller/cx_freeze?
    RESOURCES = resource_path("resources")  # root resource directory
    IMG_RESOURCES = os.path.join(RESOURCES, "img")  # where are images stored?

    DEBUG = False  # primarily controls how things are logged
    LOG_LEVEL = "info"  # debug | info | warning | error | critical
    LOG_TO_CONSOLE = False  # should logs be written to console as well as log files?

    VERSION = "v0.3.2"  # current QCPump version
    VENDOR = "QATrack Project"
    APPNAME = "qcpump"

    if frozen:
        _DEFAULT_PUMP_DIRECTORY = os.path.join(root, "qcpump", "contrib", "pumps")  # first party pump type directory
    else:
        _DEFAULT_PUMP_DIRECTORY = os.path.join(root, "contrib", "pumps")  # first party pump type directory

    PUMP_DIRECTORIES = None  # set to list of other directories to include user defined pump types from

    DB_CONNECT_TIMEOUT = 30  # timeout for database connections where available

    BROWSER_USER_AGENT = (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/70.0.3538.102 Safari/537.36 Edge/18.19582"
    )

    PUMP_ON_STARTUP = False

    def __init__(self):
        """Load settings.json file and override any settings defined in it"""

        settings_dir = Path(appdirs.user_config_dir(Settings.APPNAME, Settings.VENDOR))
        self.fname = settings_dir / "settings.json"
        if not self.fname.parent.exists():
            self.fname.parent.mkdir(parents=True, exist_ok=True)

        # if the settings.json file doesn't exist, create it and stick an empty json doc in it
        if not os.path.exists(self.fname):
            try:
                with open(self.fname, "w") as f:
                    f.write(json.dumps({
                        "LOG_LEVEL": "info",
                        "DEBUG": False,
                        "PUMP_DIRECTORIES": [],
                        "DB_CONNECT_TIMEOUT": 3,
                        "PUMP_ON_STARTUP": False,
                    }, indent=2))
            except Exception as e:
                raise Exception("Unable to create settings file %s. Error was %s" % (self.fname, str(e)))

        # now load settings file
        try:
            with open(self.fname, 'rt') as f:
                settings = json.load(f)
        except Exception as e:
            raise Exception("Settings file %s is not valid json. Error was %s" % (self.fname, str(e)))

        # override any settings user has set in settings.json
        for k, v in settings.items():
            if hasattr(self, k):
                setattr(self, k, v)

    def get_img_path(self, img_file_name):
        """return full path for image resource with filename"""
        return os.path.join(self.IMG_RESOURCES, img_file_name)

    def ico(self, icon_name):
        """return full path to icon named icon_name"""
        return os.path.join(self.IMG_RESOURCES, "icons", icon_name + ".png")
